File: backend/cqox/models/analysis.py
"""
Pydantic models for Analysis API
"""
from datetime import datetime
from typing import Optional, List, Dict, Any, Literal
from pydantic import BaseModel, Field, ConfigDict


class AnalysisBase(BaseModel):
    """Base analysis model"""
    dataset_id: int = Field(..., description="Dataset ID")
    estimator_type: str = Field(..., description="Estimator type (DR, IPW, etc.)")
    treatment_col: str = Field(..., description="Treatment column name")
    outcome_col: str = Field(..., description="Outcome column name")
    feature_cols: List[str] = Field(..., description="Feature column names")


class AnalysisCreate(AnalysisBase):
    """Analysis creation request"""
    estimators: List[str] = Field(default=['DR'], description="List of estimators to run")
    policy_id: Optional[str] = None
    scenario_spec: Optional[Dict[str, Any]] = None
    bootstrap_iterations: int = Field(default=100, description="Bootstrap iterations")
    confidence_level: float = Field(default=0.95, description="Confidence level")


class AnalysisResponse(BaseModel):
    """Analysis response model"""
    id: int = Field(alias="analysis_id")
    dataset_id: int
    user_id: int
    estimator_type: str
    treatment_col: str
    outcome_col: str
    feature_cols: List[str]
    status: Literal['pending', 'running', 'completed', 'failed']
    started_at: datetime
    completed_at: Optional[datetime] = None
    results: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None

    # Derived fields from results
    delta_yen: Optional[float] = None
    delta_yen_ci_low: Optional[float] = None
    delta_yen_ci_high: Optional[float] = None
    verdict: Optional[Literal['Go', 'Canary', 'Hold']] = None
    progress: int = 0

    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True
    )

    def __init__(self, **data):
        """Custom init to populate derived fields"""
        super().__init__(**data)

        # Extract derived fields from results
        if self.results:
            self.delta_yen = self.results.get('ate')
            self.delta_yen_ci_low = self.results.get('ci_low')
            self.delta_yen_ci_high = self.results.get('ci_high')
            self.verdict = self.results.get('verdict')

        # Calculate progress
        if self.status == 'completed':
            self.progress = 100
        elif self.status == 'running':
            self.progress = 50
        elif self.status == 'failed':
            self.progress = 0
        else:
            self.progress = 0


class AnalysisRunRequest(BaseModel):
    """Request to run causal analysis"""
    dataset_id: str
    policy_id: Optional[str] = None
    estimators: List[str] = Field(default=['DR'], description="Estimators to use")
    treatment_col: str
    outcome_col: str
    feature_cols: List[str]
    scenario_spec: Optional[Dict[str, Any]] = None
    bootstrap_iterations: int = 100
    confidence_level: float = 0.95

File: frontend/src/api/v1/analysis.ts
import { api } from '../../utils/api';

export interface AnalysisRequest {
  dataset_id: string;
  policy_id: string;
  estimators: string[];
  treatment_col: string;
  outcome_col: string;
  feature_cols: string[];
  scenario_spec: any;
  bootstrap_iterations?: number;
  confidence_level?: number;
}

export interface AnalysisStatus {
  analysis_id: string;
  policy_id: string;
  dataset_id: string;
  status: 'pending' | 'running' | 'completed' | 'failed';
  progress: number;
  delta_yen?: number;
  delta_yen_ci_low?: number;
  delta_yen_ci_high?: number;
  verdict?: 'Go' | 'Canary' | 'Hold';
  started_at?: string;
  completed_at?: string;
  error_message?: string;
  error_code?: string;
  error_details?: Record<string, any> | null;
}

export interface DiagnosticCheck {
  type: string;
  name: string;
  passed: boolean;
  score?: number;
  threshold?: number;
  data?: any;
}

export interface DiagnosticsData {
  status: string;
  cas_score: number;
  quality_level: string;
  diagnostics: DiagnosticCheck[];
  recommendations?: string[];
  total_checks: number;
}

export interface AnalysisImpactMetrics {
  estimated_cost: number;
  baseline_conversion_rate: number;
  projected_conversion_rate: number;
  conversion_uplift: number;
  users_affected: number;
  total_users: number;
  estimated_roi?: number | null;
}

export interface AnalysisDetails {
  analysis: AnalysisStatus;
  diagnostics?: DiagnosticsData | null;
  impact_metrics?: AnalysisImpactMetrics | null;
  estimator_results?: Record<string, any> | null;
}

export const analysisAPI = {
  /**
   * 因果推論分析を開始
   */
  start: async (request: AnalysisRequest): Promise<AnalysisStatus> => {
    return await api.post<AnalysisStatus>('/api/v1/analysis/run', request);
  },

  /**
   * 分析ステータス取得
   */
  getStatus: async (analysisId: string): Promise<AnalysisStatus> => {
    return await api.get<AnalysisStatus>(`/api/v1/analysis/${analysisId}`);
  },

  /**
   * 分析一覧取得
   */
  list: async (params?: {
    status?: string;
    policy_id?: string;
    page?: number;
    page_size?: number;
  }): Promise<AnalysisStatus[]> => {
    const queryParams = new URLSearchParams();
    if (params?.status) queryParams.append('status', params.status);
    if (params?.policy_id) queryParams.append('policy_id', params.policy_id);
    if (params?.page) queryParams.append('page', params.page.toString());
    if (params?.page_size) queryParams.append('page_size', params.page_size.toString());
    
    const url = `/api/v1/analysis${queryParams.toString() ? '?' + queryParams.toString() : ''}`;
    return await api.get<AnalysisStatus[]>(url);
  },

  /**
   * 分析キャンセル
   */
  cancel: async (analysisId: string): Promise<void> => {
    await api.delete(`/api/v1/analysis/${analysisId}`);
  },

  /**
   * 分析の詳細（診断結果・影響指標）取得
   */
  getDetails: async (analysisId: string): Promise<AnalysisDetails> => {
    return await api.get<AnalysisDetails>(`/api/v1/analysis/${analysisId}/details`);
  },
};
